src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/OpenSslX509CertificateReader.cs

```diff
@@ -0,0 +1,572 @@
+            callbackContext.Results.Add(new X509Extension(new Oid(oidStr), rawData, critical));
+        }
+
+        public IEnumerable<X509Extension> Extensions
+        {
+            get
+            {
+                EnumExtensionsContext context = default;
+                context.Results = new List<X509Extension>();
+                unsafe
+                {
+                    Interop.AndroidCrypto.X509EnumExtensions(_cert, &EnumExtensionsCallback, Unsafe.AsPointer(ref context));
+                }
+
+                return context.Results;
+            }
+        }
+
+        internal static ArraySegment<byte> FindFirstExtension(SafeX509Handle cert, string oidValue)
+        {
+            return Interop.AndroidCrypto.X509FindExtensionData(cert, oidValue);
+        }
+
+        internal void SetPrivateKey(SafeEvpPKeyHandle privateKey)
+        {
+            _privateKey = privateKey;
+        }
+
+        internal SafeEvpPKeyHandle? PrivateKeyHandle
+        {
+            get { return _privateKey; }
+        }
+
+        public RSA? GetRSAPrivateKey()
+        {
+            if (_privateKey == null || _privateKey.IsInvalid)
+            {
+                return null;
+            }
+
+            return new RSAOpenSsl(_privateKey);
+        }
+
+        public DSA? GetDSAPrivateKey()
+        {
+            if (_privateKey == null || _privateKey.IsInvalid)
+            {
+                return null;
+            }
+
+            return new DSAOpenSsl(_privateKey);
+        }
+
+        public ECDsa GetECDsaPublicKey()
+        {
+            using (SafeEvpPKeyHandle publicKeyHandle = Interop.Crypto.GetX509EvpPublicKey(_cert))
+            {
+                Interop.Crypto.CheckValidOpenSslHandle(publicKeyHandle);
+
+                return new ECDsaOpenSsl(publicKeyHandle);
+            }
+        }
+
+        public ECDiffieHellman GetECDiffieHellmanPublicKey()
+        {
+            using (SafeEvpPKeyHandle publicKeyHandle = Interop.Crypto.GetX509EvpPublicKey(_cert))
+            {
+                Interop.Crypto.CheckValidOpenSslHandle(publicKeyHandle);
+
+                return new ECDiffieHellmanOpenSsl(publicKeyHandle);
+            }
+        }
+
+        public ECDsa? GetECDsaPrivateKey()
+        {
+            if (_privateKey == null || _privateKey.IsInvalid)
+            {
+                return null;
+            }
+
+            return new ECDsaOpenSsl(_privateKey);
+        }
+
+        public ECDiffieHellman? GetECDiffieHellmanPrivateKey()
+        {
+            if (_privateKey == null || _privateKey.IsInvalid)
+            {
+                return null;
+            }
+
+            return new ECDiffieHellmanOpenSsl(_privateKey);
+        }
+
+        private ICertificatePal CopyWithPrivateKey(SafeEvpPKeyHandle privateKey)
+        {
+            // This could be X509Duplicate for a full clone, but since OpenSSL certificates
+            // are functionally immutable (unlike Windows ones) an UpRef is sufficient.
```

Comment:
This comment is describing an OpenSSL behavior, and doesn't seem appropriate for the Android variant.